File: src/main/java/com/yerti/core/menus/MenuListener.java
package com.yerti.core.menus;

import org.bukkit.Material;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.Inventory;

import java.util.ArrayList;
import java.util.List;

/**
 * Listener for all of the menu items
 */
public class MenuListener implements Listener {

    private List<Page> menus;
    private List<Inventory> inventories;

    /**
     * Creates a MenuListener off of a list of {@link Page}s
     *
     * @param menus
     */
    public MenuListener(List<Page> menus) {
        this.menus = menus;
        this.inventories = new ArrayList<>();
        menus.forEach(menu -> inventories.add(menu.build()));
    }

    /**
     * Checks for any specific {@link MenuItem}s and their designated events
     *
     * @param event
     */
    @EventHandler
    public void onPlayerClick(InventoryClickEvent event) {
        if (event.getInventory() == null) return;
        if (!inventories.contains(event.getInventory())) return;
        if (event.getCurrentItem() == null || event.getCurrentItem().getType().equals(Material.AIR)) return;
        event.setCancelled(true);

        if (event.getCurrentItem().getItemMeta() == null) return;

        Page menu = menus.get(inventories.indexOf(event.getInventory()));

        menu.getMenuItems().forEach(menuItem -> {
            if (event.getCurrentItem().getItemMeta().equals(menuItem.getItemMeta())) {
                if (event.getCurrentItem().getType().equals(Material.AIR)) return;
                menuItem.onClick(event);
            }
        });


    }


}

File: src/main/java/com/yerti/core/menus/MenuListener.kt
package com.yerti.core.menus

import org.bukkit.Material
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.Inventory
import java.util.*
import java.util.function.Consumer

/**
 * Listener for all of the menu items
 */
class MenuListener(private val menus: List<Page>) : Listener {
    private val inventories: MutableList<Inventory?>
    /**
     * Checks for any specific [MenuItem]s and their designated events
     * @param event
     */
    @EventHandler
    fun onPlayerClick(event: InventoryClickEvent) {
        if (event.inventory == null) return
        if (!inventories.contains(event.inventory)) return
        if (event.currentItem == null || event.currentItem.type == Material.AIR) return
        event.isCancelled = true
        if (event.currentItem.itemMeta == null) return
        val menu = menus[inventories.indexOf(event.inventory)]
        menu.menuItems.forEach(Consumer { menuItem: MenuItem? ->
            if (event.currentItem.itemMeta == menuItem!!.itemMeta) {
                if (event.currentItem.type == Material.AIR)
                menuItem.onClick(event)
            }
        })
    }

    /**
     * Creates a MenuListener off of a list of [Page]s
     * @param menus
     */
    init {
        inventories = ArrayList()
        menus.forEach(Consumer { menu: Page -> inventories.add(menu.build()) })
    }
}